crates/telio-traversal/src/session_keeper.rs

```diff
@@ -111,113 +114,143 @@ impl SessionKeeper {
+            })
+            .await
+            .map_err(Error::Task)?,
 
-                                let secondary_client = match second {
-                                    IpAddr::V4(_) => &c.pinger_client_v4,
-                                    IpAddr::V6(_) => &c.pinger_client_v6,
-                                };
+            None => task_exec!(&self.task, async move |s| {
+                if s.nonbatched_actions.contains_action(&public_key) {
+                    let _ = s.nonbatched_actions.remove_action(&public_key);
+                }
 
-                                let _ = secondary_client
-                                    .pinger(second, PingIdentifier(rand::random()))
-                                    .await
-                                    .send_ping(PingSequence(0), &[0; PING_PAYLOAD_SIZE])
-                                    .await?;
+                Ok(s.nonbatched_actions.add_action(
+                    public_key,
+                    interval,
+                    Arc::new(move |c| {
+                        Box::pin(async move {
+                            if let Err(e) = ping(&c.pingers, (&public_key, &dual_target)).await {
+                                telio_log_warn!(
+                                    "Failed to ping, peer with key: {:?}, error: {:?}",
+                                    public_key,
+                                    e
+                                );
                             }
-                        }
-                        Ok(())
-                    })
-                }),
-            ))
-        })
-        .await
-        .map_err(Error::Task)?
-        .map_err(Error::RepeatedActionError)
-        .map(|_| ())
-    }
+                            Ok(())
+                        })
+                    }),
+                ))
+            })
+            .await
+            .map_err(Error::Task)?
+            .map_err(Error::RepeatedActionError)
+            .map(|_| ())?,
+        }
 
-    async fn update_interval(&self, public_key: &PublicKey, interval: Duration) -> Result<()> {
-        let public_key = *public_key;
-        task_exec!(&self.task, async move |s| {
-            let _ = s.keepalive_actions.update_interval(&public_key, interval);
-            Ok(())
-        })
-        .await
-        .map_err(Error::Task)
+        Ok(())
     }
 
-    async fn remove_node(&self, public_key: &PublicKey) -> Result<()> {
-        // FIXME: error handling with task_exec! seems to suck a lot. Need to fix that.
-        let public_key = *public_key;
+    // TODO: this only removes the batched key. There's probably a leak in the current implementation where the entries are not cleaned-up, only added.
+    async fn remove_node(&self, key: &PublicKey) -> Result<()> {
+        let key = *key;
         task_exec!(&self.task, async move |s| {
-            let _ = s.keepalive_actions.remove_action(&public_key);
+            let _ = s.batched_actions.remove(&key);
```

Comment:
why don't we do here `s.nonbatched_actions.remove_action(&key)` also?

Reply:
Hmmm, my bad. I deliberately added `TODO` on top to address this while I left a bug. Thanks for picking up

Reply:
fixed it